lib/fog/azurerm/models/storage/file.rb

```diff
@@ -4,147 +4,386 @@ class AzureRM
+              break if data.nil?
+              block_id = Base64.strict_encode64(random_string(32))
+              service.put_blob_block(directory.key, key, block_id, data, options)
+              blocks << [block_id]
+            end
+          else
+            body.bytes.each_slice(MAXIMUM_CHUNK_SIZE) do |chunk|
+              block_id = Base64.strict_encode64(random_string(32))
+              service.put_blob_block(directory.key, key, block_id, chunk.pack('C*'), options)
+              blocks << [block_id]
+            end
+          end
+
+          # Complete the upload
+          service.commit_blob_blocks(directory.key, key, blocks, options)
+        rescue
+          # Abort the upload & reraise
+          begin
+            service.delete_blob(directory.key, key)
+          rescue
+            nil
+          end
+          raise
         end
 
-        def self.parse_object(blob)
-          hash = {}
-          hash['key'] = blob.name
-          hash['metadata'] = blob.metadata
-          return hash unless blob.respond_to?('properties')
-
-          hash['last_modified'] = blob.properties[:last_modified]
-          hash['etag'] = blob.properties[:etag]
-          hash['lease_duration'] = blob.properties[:lease_duration]
-          hash['lease_status'] = blob.properties[:lease_status]
-          hash['lease_state'] = blob.properties[:lease_state]
-          hash['content_length'] = blob.properties[:content_length]
-          hash['content_type'] = blob.properties[:content_type]
-          hash['content_encoding'] = blob.properties[:content_encoding]
-          hash['content_language'] = blob.properties[:content_language]
-          hash['content_disposition'] = blob.properties[:content_disposition]
-          hash['content_md5'] = blob.properties[:content_md5]
-          hash['cache_control'] = blob.properties[:cache_control]
-          hash['sequence_number'] = blob.properties[:sequence_number]
-          hash['blob_type'] = blob.properties[:blob_type]
-          hash['copy_id'] = blob.properties[:copy_id]
-          hash['copy_status'] = blob.properties[:copy_status]
-          hash['copy_source'] = blob.properties[:copy_source]
-          hash['copy_progress'] = blob.properties[:copy_progress]
-          hash['copy_completion_time'] = blob.properties[:copy_completion_time]
-          hash['copy_status_description'] = blob.properties[:copy_status_description]
-          hash['accept_ranges'] = blob.properties[:accept_ranges]
-          hash
+        # Upload a page blob
+        def save_page_blob(blob_size, options)
+          # Initiate the upload
+          service.create_page_blob(directory.key, key, blob_size, options)
+
+          # Uploading content
+          # TODO: Upload content in parallel using threads
+          if body.respond_to?(:read)
+            if body.respond_to?(:rewind)
+              begin
+                body.rewind
+              rescue
+                nil
```

Comment:
If body does not support rewind, we can ignore it. Reference 'https://github.com/fog/fog-aws/blob/master/lib/fog/aws/models/storage/file.rb#L272'